src/components/menu/menuOpener/MenuOpener.tsx

```diff
@@ -0,0 +1,61 @@
+// (C) 2007-2018 GoodData Corporation
+import * as React from 'react';
+import { OpenAction, IMenuPositionConfig } from '../MenuSharedTypes';
+import MenuOpenedByClick from './MenuOpenedByClick';
+import MenuOpenedByHover from './MenuOpenedByHover';
+
+export interface IMenuOpenerProps extends Partial<IMenuPositionConfig> {
+    topLevel: boolean;
+    opened: boolean;
+    onOpenedChange: (opened: boolean) => void;
+    openAction?: OpenAction;
+    portalTarget?: Element;
+    toggler: React.ReactNode;
+    children: React.ReactNode;
+}
+
+export default class MenuOpener extends React.Component<IMenuOpenerProps> {
+    public static defaultProps: Partial<IMenuOpenerProps> = {
+        openAction: 'hover',
+
+        alignment: ['right', 'bottom'],
+        spacing: 0,
+        offset: 0,
+
+        portalTarget: document.querySelector('body')
+    };
+
+    public render() {
+        let Component;
+        switch (this.props.openAction) {
```

Comment:
please extract this 29-36 to separate method

Reply:
Moved